business/apps.go

```diff
@@ -48,13 +48,75 @@ func buildFinalLabels(m map[string][]string) map[string]string {
 }
 
 // GetAppList is the API handler to fetch the list of applications in a given namespace
-func (in *AppService) GetAppList(namespace string) (models.AppList, error) {
+func (in *AppService) GetAppList(namespace string, linkIstioResources bool) (models.AppList, error) {
 	appList := &models.AppList{
 		Namespace: models.Namespace{Name: namespace},
 		Apps:      []models.AppListItem{},
 	}
-	apps, err := fetchNamespaceApps(in.businessLayer, namespace, "")
-	if err != nil {
+
+	var err error
+	var apps namespaceApps
+
+	var virtualServices []kubernetes.IstioObject
+	var destinationRules []kubernetes.IstioObject
+	var gateways []kubernetes.IstioObject
+	var authorizationPolicies []kubernetes.IstioObject
+	var peerAuthentications []kubernetes.IstioObject
+	var sidecars []kubernetes.IstioObject
+	var requestAuthentications []kubernetes.IstioObject
+	var envoyFilters []kubernetes.IstioObject
+
+	nFetches := 1
+	if linkIstioResources {
+		nFetches = 9
+	}
+
+	wg := sync.WaitGroup{}
+	wg.Add(nFetches)
+	errChan := make(chan error, nFetches)
+
+	go func() {
+		defer wg.Done()
+		var err2 error
+		apps, err2 = fetchNamespaceApps(in.businessLayer, namespace, "")
+		if err2 != nil {
+			log.Errorf("Error fetching Applications per namespace %s: %s", namespace, err2)
+			errChan <- err2
+		}
+	}()
+
+	linkedResources := map[string]*[]kubernetes.IstioObject{
+		kubernetes.VirtualServices:        &virtualServices,
+		kubernetes.DestinationRules:       &destinationRules,
+		kubernetes.Gateways:               &gateways,
+		kubernetes.AuthorizationPolicies:  &authorizationPolicies,
+		kubernetes.PeerAuthentications:    &peerAuthentications,
+		kubernetes.Sidecars:               &sidecars,
+		kubernetes.RequestAuthentications: &requestAuthentications,
+		kubernetes.EnvoyFilters:           &envoyFilters,
+	}
+
+	if linkIstioResources {
+		for linkedResource, dest := range linkedResources {
+			go func(namespace, resourceType string, dest *[]kubernetes.IstioObject, errChan chan error) {
+				defer wg.Done()
+				var err2 error
+				if IsNamespaceCached(namespace) {
+					*dest, err2 = kialiCache.GetIstioObjects(namespace, resourceType, "")
+				} else {
+					*dest, err2 = in.k8s.GetIstioObjects(namespace, resourceType, "")
+				}
+				if err2 != nil {
+					log.Errorf("Error fetching Istio %s per namespace %s: %s", resourceType, namespace, err2)
+					errChan <- err2
+				}
+			}(namespace, linkedResource, dest, errChan)
+		}
+	}
```

Comment:
Sorry there is still one thing bothering me a bit. I thought the individual declaration on lines L60-L67 were necessaries to avoid race conditions. But I guess it is not. So you can just fill a hashmap with all the resources by resource name.

```go
	resources := []string{ kubernetes.VirtualServices, kubernetes.DestinationRules, kubernetes.Gateways,
		kubernetes.AuthorizationPolicies, kubernetes.PeerAuthentications, kubernetes.Sidecars,
		kubernetes.RequestAuthentications, kubernetes.EnvoyFilters,
	}

	linkedResources := map[string]*[]kubernetes.IstioObject{}
	if linkIstioResources {
		for _, resource := range resources {
			var resourceObjects []kubernetes.IstioObject
			linkedResources[resource] = &resourceObjects
			go func(namespace, resourceType string, dest *[]kubernetes.IstioObject, errChan chan error) {
				defer wg.Done()
				var err2 error
				if IsNamespaceCached(namespace) {
					*dest, err2 = kialiCache.GetIstioObjects(namespace, resourceType, "")
				} else {
					*dest, err2 = in.k8s.GetIstioObjects(namespace, resourceType, "")
				}
				if err2 != nil {
					log.Errorf("Error fetching Istio %s per namespace %s: %s", resourceType, namespace, err2)
					errChan <- err2
				}
			}(namespace, resource, &resourceObjects, errChan)
		}
	}
```

With this you can get rid of the declaration of one var for each resource type. Adding one more resource will be as easy as adding the type in the `resources` array (L60-L67).

Reply:
Oh, true, there are some intermediate variables that are not needed.
Thanks